beaker_kernel/lib/context.py

```diff
@@ -76,14 +80,49 @@ def __init__(self, beaker_kernel: "LLMKernel", subkernel: "BaseSubkernel", agent
     async def setup(self, config=None, parent_header=None):
         if config:
             self.config = config
+
         if callable(getattr(self.agent, 'setup', None)):
             await self.agent.setup(self.config, parent_header=parent_header)
 
     async def cleanup(self):
+        self.subkernel.cleanup()
         for msg_type, intercept_func, stream in self.intercepts:
             self.beaker_kernel.remove_intercept(msg_type=msg_type, func=intercept_func, stream=stream)
         del self.agent
 
+    def _collect_and_register_intercepts(self, target):
+        for _, method in inspect.getmembers(target, lambda member: inspect.ismethod(member) and hasattr(member, "_intercept")):
+            msg_type, stream = getattr(method, "_intercept")
+            self.intercepts.append((msg_type, method, stream))
+            self.beaker_kernel.add_intercept(msg_type=msg_type, func=method, stream=stream)
+
+    def _init_subkernel(self):
+        language = self.config.get("language", "python3")
+        self.beaker_kernel.debug("new_kernel", f"Setting new kernel of `{language}`")
+        kernel_opts = {
+            subkernel.KERNEL_NAME: subkernel
+            for subkernel in autodiscover("subkernels").values()
+        }
+        res = requests.post(
+            f"{server_url}/api/kernels",
+            json={"name": language, "path": ""},
+            headers={"Authorization": f"token {server_token}"},
+        )
+        kernel_info = res.json()
+        self.beaker_kernel.update_running_kernels()
+        kernels = self.beaker_kernel.kernels
+        subkernel_id = kernel_info["id"]
+        # NOTE: MODIFIED `connect_to`
+        # TODO: Refactor this into `lib/kernel_proxy_manager.py`
+        matching = next((n for n in kernels if subkernel_id in n), None)
+        if matching is None:
+            raise ValueError("Unknown kernel " + subkernel_id)
+        if kernels[matching] == self.beaker_kernel.server.config:
+            raise ValueError("Refusing loopback connection")
+        self.subkernel = kernel_opts[language](subkernel_id, kernels[matching], self)
+        self.beaker_kernel.server.set_proxy_target(self.subkernel.connected_kernel)
```

Comment:
Couldn't this just be: `self.beaker_kernel.connect_to(subkernel_id)`?

Reply:
following our slack discussion, we concluded that this will actually create a duplicate kernel. resolving without changing for now